File: packages/runtime/src/builtin.ts
import * as tg from "./index.ts";

export type ArchiveFormat = "tar" | "tgar" | "zip";

export type CompressionFormat = "bz2" | "gz" | "xz" | "zst";

export type DownloadOptions = {
	decompress?: boolean;
	extract?: boolean;
};

export let archive = async (
	artifact: tg.Artifact,
	format: ArchiveFormat,
	compression?: tg.CompressionFormat,
): Promise<tg.Blob> => {
	let value = await tg.build({
		args: [artifact, format, compression],
		executable: "archive",
		host: "builtin",
	});
	tg.assert(tg.Blob.is(value));
	return value;
};

export let bundle = async (artifact: tg.Artifact): Promise<tg.Artifact> => {
	let value = await tg.build({
		args: [artifact],
		executable: "bundle",
		host: "builtin",
	});
	tg.assert(tg.Artifact.is(value));
	return value;
};

export let checksum = async (
	input: tg.Blob | tg.Artifact,
	algorithm: tg.Checksum.Algorithm,
): Promise<tg.Checksum> => {
	let value = await tg.build({
		args: [input, algorithm],
		executable: "checksum",
		host: "builtin",
	});
	return value as tg.Checksum;
};

export let compress = async (
	blob: tg.Blob,
	format: tg.CompressionFormat,
): Promise<tg.Blob> => {
	let value = await tg.build({
		args: [blob, format],
		executable: "compress",
		host: "builtin",
	});
	tg.assert(tg.Blob.is(value));
	return value;
};

export let decompress = async (blob: tg.Blob): Promise<tg.Blob> => {
	let value = await tg.build({
		args: [blob],
		executable: "decompress",
		host: "builtin",
	});
	tg.assert(tg.Blob.is(value));
	return value;
};

export let download = async (
	url: string,
	checksum: tg.Checksum,
	options?: DownloadOptions,
): Promise<tg.Blob | tg.Artifact> => {
	let value = await tg.build({
		args: [url, options ?? { mode: "raw" }],
		checksum,
		executable: "download",
		host: "builtin",
	});
	tg.assert(tg.Blob.is(value) || tg.Artifact.is(value));
	return value;
};

export let extract = async (blob: tg.Blob): Promise<tg.Artifact> => {
	let value = await tg.build({
		args: [blob],
		executable: "extract",
		host: "builtin",
	});
	tg.assert(tg.Artifact.is(value));
	return value;
};

File: packages/clients/python/tangram/builtin.py
"""Builtin functions matching JS builtin.ts."""

from typing import Any, Literal, Optional, TypedDict, Union

ArchiveFormat = Literal["tar", "tgar", "zip"]
CompressionFormat = Literal["bz2", "gz", "xz", "zst"]
ChecksumAlgorithm = Literal["blake3", "sha256", "sha512"]


class DownloadOptions(TypedDict, total=False):
    """Options for the download function."""

    checksum: Optional[ChecksumAlgorithm]
    mode: Optional[Literal["raw", "decompress", "extract"]]


async def archive(
    artifact: Any,
    format: ArchiveFormat,
    compression: Optional[CompressionFormat] = None,
) -> Any:
    """Create an archive from an artifact.

    Args:
        artifact: The artifact to archive.
        format: The archive format (tar, tgar, zip).
        compression: Optional compression format.

    Returns:
        A blob containing the archive.
    """
    from tangram.build import build
    from tangram.blob import Blob

    value = await build(
        host="builtin",
        executable="archive",
        args=[artifact, format, compression],
    )
    assert isinstance(value, Blob)
    return value


async def bundle(artifact: Any) -> Any:
    """Bundle an artifact.

    Args:
        artifact: The artifact to bundle.

    Returns:
        The bundled artifact.
    """
    from tangram.build import build

    value = await build(
        host="builtin",
        executable="bundle",
        args=[artifact],
    )
    return value


async def checksum(
    input_: Any,
    algorithm: ChecksumAlgorithm,
) -> str:
    """Compute a checksum.

    Args:
        input_: The blob or artifact to checksum.
        algorithm: The checksum algorithm.

    Returns:
        The checksum string.
    """
    from tangram.build import build

    value = await build(
        host="builtin",
        executable="checksum",
        args=[input_, algorithm],
    )
    assert isinstance(value, str)
    return value


async def compress(
    blob: Any,
    format: CompressionFormat,
) -> Any:
    """Compress a blob.

    Args:
        blob: The blob to compress.
        format: The compression format.

    Returns:
        The compressed blob.
    """
    from tangram.build import build
    from tangram.blob import Blob

    value = await build(
        host="builtin",
        executable="compress",
        args=[blob, format],
    )
    assert isinstance(value, Blob)
    return value


async def decompress(blob: Any) -> Any:
    """Decompress a blob.

    Args:
        blob: The blob to decompress.

    Returns:
        The decompressed blob.
    """
    from tangram.build import build
    from tangram.blob import Blob

    value = await build(
        host="builtin",
        executable="decompress",
        args=[blob],
    )
    assert isinstance(value, Blob)
    return value


async def download(
    url: str,
    checksum_value: str,
    options: Optional[DownloadOptions] = None,
) -> Any:
    """Download a file from a URL.

    Args:
        url: The URL to download from.
        checksum_value: The expected checksum of the file.
        options: Optional download options.

    Returns:
        The downloaded blob or artifact.
    """
    from tangram.build import build

    if options is None:
        options = {}

    value = await build(
        host="builtin",
        executable="download",
        args=[url, options],
        checksum=checksum_value,
    )
    return value


async def extract(blob: Any) -> Any:
    """Extract an archive.

    Args:
        blob: The archive blob to extract.

    Returns:
        The extracted artifact.
    """
    from tangram.build import build

    value = await build(
        host="builtin",
        executable="extract",
        args=[blob],
    )
    return value
